### ISSUE TITLE
X.509 trusted certificates location

### ISSUE THREAD

**Issue by rowlap** [FreeRDP/FreeRDP#6989]
**Describe the bug**
Similar to issue #2446, there's a problem connecting to Windows 10 when using an internal trusted root certificate.

The root cert has been added to the system [in accordance with the distro requirements](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/security_guide/sec-shared-system-certificates) however it appears xfreerdp is bypassing the default file path, checking only the hash-based names which `update-ca-trust` doesn't create.

**Steps to reproduce**
1. Set up internal PKI with root > intermediate > end certificate
2. On any Red Hat-derived distro, place the internal trusted root certificate under /etc/pki/ca-trust/source/anchors/
3. Run `update-ca-trust` to populate the well-known locations
4. Verify that `trust list` reports the internal CA
5. Ensure that `openssl s_client -connect HOSTNAME:3389` succeeds, in as far as connecting without verification errors
6. Attempt to connect with `xfreerdp /v:HOSTNAME /from-stdin`

**What's expected?**
Prompt for credentials.

**What happens?**
Certificate verification failure.

```
$ strace -f -e trace=file xfreerdp /v:win10-example /from-stdin
execve("/usr/bin/xfreerdp", ["xfreerdp", "/v:win10-example", "/from-stdin"], 0x7ffd348c7638 /* 42 vars */) = 0
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
# skip to pki files

access("/etc/selinux/config", F_OK)     = 0
access("/etc/system-fips", F_OK)        = -1 ENOENT (No such file or directory)
open("/etc/pki/tls/legacy-settings", O_RDONLY) = -1 ENOENT (No such file or directory)
access("/etc/system-fips", F_OK)        = -1 ENOENT (No such file or directory)
open("/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
open("/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib/locale/C.utf8/LC_COLLATE", O_RDONLY|O_CLOEXEC) = 3
open("/usr/lib64/gconv/gconv-modules.cache", O_RDONLY) = 3
open("/etc/winpr/HKLM.reg", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("/etc/winpr/HKLM.reg", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("/etc/winpr/HKLM.reg", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("/etc/winpr/HKLM.reg", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("/dev/urandom", O_RDONLY|O_NOCTTY|O_NONBLOCK) = 5
access("/home/phr/.Xauthority", R_OK) = 0
open("/home/phr/.Xauthority", O_RDONLY) = 11

# further skip to pki files
[pid 30523] stat("/home/phr/.config/freerdp", {st_mode=S_IFDIR|0775, st_size=77, ...}) = 0
[pid 30523] stat("/home/phr/.config/freerdp/certs", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
[pid 30523] stat("/home/phr/.config/freerdp/server", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
[pid 30523] stat("/etc/pki/tls/certs/4d2b349a.0", 0x7f28be7ff370) = -1 ENOENT (No such file or directory)
[pid 30523] stat("/home/phr/.config/freerdp/certs/4d2b349a.0", 0x7f28be7ff370) = -1 ENOENT (No such file or directory)
[pid 30523] stat("/etc/pki/tls/certs/7aa2e51f.0", 0x7f28be7ff370) = -1 ENOENT (No such file or directory)
[pid 30523] stat("/home/phr/.config/freerdp/certs/7aa2e51f.0", 0x7f28be7ff370) = -1 ENOENT (No such file or directory)
[15:12:55:219] [30522:30523] [WARN][com.freerdp.crypto] - Certificate verification failure 'unable to get local issuer certificate (20)' at stack position 1
[15:12:55:219] [30522:30523] [WARN][com.freerdp.crypto] - DC = com, DC = EXAMPLE, CN = Example CA1

Certificate details for win10-example:3389 (RDP-Server):
        Common Name: WIN10-EXAMPLE
        Subject:     DC = com, DC = EXAMPLE, OU = EX, OU = Workstations, CN = WIN10-EXAMPLE
        Issuer:      DC = com, DC = EXAMPLE, CN = Example CA1
        Thumbprint:  xx:xx:xx:xx:xx:xx
The above X.509 certificate could not be verified, possibly because you do not have
the CA certificate in your certificate store, or the certificate has expired.
Please look at the OpenSSL documentation on how to add a private CA to the store.
[15:12:55:229] [30522:30523] [ERROR][com.freerdp.crypto] - certificate not trusted, aborting.
[15:12:55:230] [30522:30523] [ERROR][com.freerdp.core] - transport_connect_tls:freerdp_set_last_error_ex ERRCONNECT_TLS_CONNECT_FAILED [0x00020008]
```

**Application details**
* FreeRDP version 2.2.0
* Command line used: `xfreerdp /v:win10-example /from-stdin`
* OS version connecting to (server side): Windows 10

**Environment**
 - OS: Linux
 - Version/Distribution: CentOS 7
 - Architecture: x86_64

**Additional context**
A workaround is to populate the /etc/pki/tls/certs directory with the hash-named symlinks, but this additional step just for xfreerdp is burdensome and unintuitive (when other openssl apps don't require it).

How can we persuade xfreerdp to consult the "bundle" file at its default location during certificate verification?

**Comment by akallabeth** [comment]
@rowlap we add `X509_LOOKUP_file` and `X509_LOOKUP_hash_dir` as store lookup locations, which one do you require additionally?

**Comment by akallabeth** [comment]
@rowlap found a possible issue, could you check #6991 fixes your issue?

**Comment by rowlap** [comment]
Looks good to me. I now see `/etc/pki/tls/cert.pem` (the CentOS 7 default file) being loaded:
```
[pid  5007] stat("/home/phr/.config/freerdp", {st_mode=S_IFDIR|0775, st_size=77, ...}) = 0
[pid  5007] stat("/home/phr/.config/freerdp/certs", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
[pid  5007] stat("/home/phr/.config/freerdp/server", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
[pid  5007] open("/etc/pki/tls/cert.pem", O_RDONLY) = 34
```

however if the bundle file doesn't exist, then the hash-based names are then checked as before:
```
[pid  5431] stat("/home/phr/.config/freerdp", {st_mode=S_IFDIR|0775, st_size=77, ...}) = 0
[pid  5431] stat("/home/phr/.config/freerdp/certs", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
[pid  5431] stat("/home/phr/.config/freerdp/server", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
[pid  5431] open("/etc/pki/tls/cert.pem", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid  5431] stat("/etc/pki/tls/certs/4d3a934d.0", 0x7f131f838370) = -1 ENOENT (No such file or directory)
[pid  5431] stat("/home/phr/.config/freerdp/certs/4d3a934d.0", 0x7f131f838370) = -1 ENOENT (No such file or directory)
[pid  5431] stat("/etc/pki/tls/certs/7cc15539.0", 0x7f131f838370) = -1 ENOENT (No such file or directory)
[pid  5431] stat("/home/phr/.config/freerdp/certs/7cc15539.0", 0x7f131f838370) = -1 ENOENT (No such file or directory)
[10:11:07:936] [5430:5431] [WARN][com.freerdp.crypto] - Certificate verification failure 'unable to get local issuer certificate (20)' at stack position 1
```

Setting `SSL_CERT_FILE` and/or `SSL_CERT_DIR` in the environment has the desired effect of changing openssl's "default" location at runtime. I appreciate freerdp goes further in managing its own certs directory, but for us the centralized file works best as end-users don't have to think about it.

**Comment by akallabeth** [comment]
@rowlap ok, so the main issue is fixed?
As for the custom cert path, that is desired behavior for the use cases we had up until now. If you want to change that pull requests are welcome (maybe configurable by some config in `/etc`?)

**Comment by rowlap** [comment]
Yes, the issue is fixed with #6991. I don't object to the `~/.freerdp/` paths, we just don't use them.
